### ISSUE TITLE
Is HPyModuleDef->globals used anywhere?

### ISSUE THREAD

**Issue by mattip** [hpyproject/hpy#462]
I don't see tests for setting the `globals` field in `HPyModuleDef`, and the only reference I could find in the cpython backend is [here](https://github.com/hpyproject/hpy/blob/a5c70cd1c19304bb2a164ab3dd69ab46b1ed542a/hpy/devel/src/runtime/ctx_module.c#L116) for error checking. 

The NumPy port does [use this](https://github.com/hpyproject/numpy-hpy/blob/23f4a1898d37af3dd3525144eb800c64abfdb495/numpy/core/src/multiarray/multiarraymodule.c#L6936), but I think that is redundant, since there the objects are globally saved and retrieved but not from the module. Using `.global` would require being able to fish out the module handles, which is discussed elsewhere.

**Comment by fangerer** [comment]
You are right. The idea of `HPyModuleDef.globals` is that one needs to register all `HPyGlobal` variables such that the interpreter can initialize the C-level value of those and do any internal preparation to be able to use the globals.

For CPython, we don't actually need that because `HPyGlobal` variables are just C global variables of type `PyObject *` and those are guaranteed to be initialized to `NULL`.

But I agree with you: since the contract is that one *must* register the globals, we should enforce this in the debug mode.
We could do that by writing some marker value to the `HPyGlobal`.

**Comment by steve-s** [comment]
I though I already had an issue opened for that. I took the liberty and renamed the issue. I think we are using that field in GraalPy (or at least planned to use it) to preallocate our internal array with the actual values for globals.

**Comment by fangerer** [comment]
yes, we already do that: https://github.com/oracle/graalpython/blob/7279f9c3ee0b9ea6a2bef578af88f575f9ed9f31/graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/objects/cext/hpy/GraalHPyNodes.java#L490
